src/main/scala/is/hail/annotations/MemoryBlock.scala

```diff
@@ -9,147 +10,143 @@ import is.hail.utils._
+    Memory.loadAddress(mem, off)
   }
 
   def loadByte(off: Long): Byte = {
-    assert(size <= length)
+    assert(size <= capacity)
     assert(off >= 0 && off + 1 <= size)
-    Memory.loadByte(mem + off)
+    Memory.loadByte(mem, off)
   }
 
   def loadBytes(off: Long, n: Int): Array[Byte] = {
-    assert(size <= length)
+    assert(size <= capacity)
     assert(off >= 0 && off + n <= size)
     val a = new Array[Byte](n)
-    Memory.copyToArray(a, 0, mem + off, n)
+    Memory.memcpy(a, 0, mem, off, n)
     a
   }
 
-  def loadBytes(off: Long, n: Long, dst: Array[Byte]) {
-    assert(size <= length)
+  def loadBytes(off: Long, to: Array[Byte], toOff: Long, n: Long) {
+    assert(size <= capacity)
     assert(off >= 0 && off + n <= size)
-    assert(n <= dst.length)
-    Memory.copyToArray(dst, 0, mem + off, n)
+    assert(toOff + n <= to.length)
+    Memory.memcpy(to, toOff, mem, off, n)
   }
 
   def storeInt(off: Long, i: Int) {
-    assert(size <= length)
+    assert(size <= capacity)
     assert(off >= 0 && off + 4 <= size)
-    Memory.storeInt(mem + off, i)
+    Memory.storeInt(mem, off, i)
   }
 
   def storeLong(off: Long, l: Long) {
-    assert(size <= length)
+    assert(size <= capacity)
     assert(off >= 0 && off + 8 <= size)
-    Memory.storeLong(mem + off, l)
+    Memory.storeLong(mem, off, l)
   }
 
   def storeFloat(off: Long, f: Float) {
-    assert(size <= length)
+    assert(size <= capacity)
     assert(off >= 0 && off + 4 <= size)
-    Memory.storeFloat(mem + off, f)
+    Memory.storeFloat(mem, off, f)
   }
 
   def storeDouble(off: Long, d: Double) {
-    assert(size <= length)
+    assert(size <= capacity)
     assert(off >= 0 && off + 8 <= size)
-    Memory.storeDouble(mem + off, d)
+    Memory.storeDouble(mem, off, d)
   }
 
   def storeAddress(off: Long, a: Long) {
-    assert(size <= length)
+    assert(size <= capacity)
     assert(off >= 0 && off + 8 <= size)
-    Memory.storeAddress(mem + off, a)
+    Memory.storeAddress(mem, off, a)
   }
 
   def storeByte(off: Long, b: Byte) {
-    assert(size <= length)
+    assert(size <= capacity)
     assert(off >= 0 && off + 1 <= size)
-    Memory.storeByte(mem + off, b)
+    Memory.storeByte(mem, off, b)
   }
 
   def storeBytes(off: Long, bytes: Array[Byte]) {
     storeBytes(off, bytes, 0, bytes.length)
   }
 
   def storeBytes(off: Long, bytes: Array[Byte], bytesOff: Long, n: Int) {
-    assert(size <= length)
+    assert(size <= capacity)
     assert(off >= 0 && off + n <= size)
     assert(bytesOff + n <= bytes.length)
-    Memory.copyFromArray(mem + off, bytes, bytesOff, n)
+    Memory.memcpy(mem, off, bytes, bytesOff, n)
   }
 
   def ensure(n: Long) {
     val required = size + n
-    if (length < required) {
-      val newLength = (length * 2).max(required)
-      mem = Memory.realloc(mem, newLength)
-      length = newLength
+    if (capacity < required) {
+      val newLength = (capacity * 2).max(required)
+      mem = util.Arrays.copyOf(mem, newLength.toInt)
```

Comment:
I added asserts before all calls to `toInt`.  I think the max is overkill here.

In practice, I don't this will be an issue for some time.  If we're creating 2GB+ regions, we're going to have other problems.  By the time we get to that point, we should be using native allocation and 64-bit pointers again.